### PR TITLE
[libc] Bound the worst-case stack usage in qsort().

### PR DESCRIPTION
**Pull request by statham-arm**
Previously, the Quicksort implementation was written in the obvious way: after each partitioning step, it explicitly recursed twice to sort the two sublists. Now it compares the two sublists' sizes, and recurses only to sort the smaller one. To handle the larger list it loops back round to the top of the function, so as to handle it within the existing stack frame.

This means that every recursive call is handling a list at most half that of its caller. So the maximum recursive call depth is O(lg N). Otherwise, in Quicksort's bad cases where each partition step peels off a small constant number of array elements, the stack usage could grow linearly with the array being sorted, i.e. it might be Θ(N).

I tested this code by manually constructing a List Of Doom that causes this particular quicksort implementation to hit its worst case, and confirming that it recursed very deeply in the old code and doesn't in the new code. But I haven't added that list to the test suite, because the List Of Doom has to be constructed in a way based on every detail of the quicksort algorithm (pivot choice and partitioning strategy), so it would silently stop being a useful regression test as soon as any detail changed.

### CODE REVIEW

File: libc/src/stdlib/quick_sort.h [hunk 1]
```diff
@@ -59,17 +59,33 @@ static size_t partition(const Array &array) {
...
+    // Make Arrays describing the two sublists that still need sorting.
+    Array left = array.make_array(0, split_index);
+    Array right = array.make_array(split_index, array.size() - split_index);
+
+    // Recurse to sort the smaller of the two, and then loop round within this
+    // function to sort the larger. This way, recursive call depth is bounded
+    // by log2 of the total array size, because every recursive call is sorting
+    // a list at most half the length of the one in its caller.
+    if (left.size() < right.size()) {
+      quick_sort(left, depth + 1);
```
Comment (smithp35) [inline libc/src/stdlib/quick_sort.h:82]: Was the depth part of your original test to check the maximum depth. Looks like there isn't a quick_sort() that takes 2 parameters.

File: libc/src/stdlib/quick_sort.h [hunk 2]
```diff
@@ -59,17 +59,33 @@ static size_t partition(const Array &array) {
...
+LIBC_INLINE void quick_sort(Array array) {
+  while (true) {
+    const size_t array_size = array.size();
+    if (array_size <= 1)
+      return;
+    size_t split_index = partition(array);
+    if (array_size <= 2) {
+      // The partition operation sorts the two element array.
+      return;
+    }
```
Comment (nickdesaulniers) [inline libc/src/stdlib/quick_sort.h:71]: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

File: libc/src/stdlib/quick_sort.h (hunk 1 above)
Comment (statham-arm) [reply libc/src/stdlib/quick_sort.h:82]: Yikes! Yes, that's a piece of diagnostic code that escaped into the wild.

(I'm happy to share the diagnostic code in a more organised way if anyone does want it, but I hadn't intended to share a small fragment of it here :-)

File: libc/src/stdlib/quick_sort.h (hunk 2 above)
Comment (statham-arm) [reply libc/src/stdlib/quick_sort.h:71]: That construction was unchanged from the previous version of the code – all _I've_ done is to move it right by two spaces! But fine, I can fix it while I'm here.

File: libc/src/stdlib/quick_sort.h [hunk 3]
```diff
@@ -59,17 +59,32 @@ static size_t partition(const Array &array) {
...
-  if (array_size <= 2) {
-    // The partition operation sorts the two element array.
-    return;
+LIBC_INLINE void quick_sort(Array array) {
+  while (true) {
+    const size_t array_size = array.size();
+    if (array_size <= 1)
+      return;
+    size_t split_index = partition(array);
+    if (array_size <= 2)
```
Comment (nickdesaulniers) [inline libc/src/stdlib/quick_sort.h:68]:
```suggestion
    if (array_size == 2)
```

File: libc/src/stdlib/qsort_data.h [hunk 4]
```diff
@@ -92,6 +92,12 @@ class Array {
   Array make_array(size_t i, size_t s) const {
     return Array(get(i), s, elem_size, compare);
   }
+
+  // Reset this Array to point at a different interval of the same items.
+  void reset_bounds(uint8_t *a, size_t s) {
```
Comment (nickdesaulniers) [inline libc/src/stdlib/qsort_data.h:97]: This method, and the others should probably have the `LIBC_INLINE` macro added.
https://libc.llvm.org/dev/code_style.html#inline-functions-and-variables-defined-in-header-files

**Review by nickdesaulniers — APPROVED** [review]

**Review by michaelrj-google — APPROVED** [review]
LGTM, if possible please also file a bug for further qsort cleanup, specifically marking all the functions in headers with `LIBC_INLINE`.

**Pull request merged**